troubleshooting/checkpointing/src/solution/java/com/ververica/flink/training/solutions/CheckpointingJobSolution3.java

```diff
@@ -0,0 +1,297 @@
+								.<FakeKafkaRecord>forBoundedOutOfOrderness(Duration.ofMillis(250))
+								.withTimestampAssigner(
+										(element, timestamp) -> element.getTimestamp())
+								.withIdleness(Duration.ofSeconds(1)))
+				.name("Watermarks")
+				.uid("Watermarks")
+				.flatMap(new MeasurementDeserializer())
+				.name("Deserialization")
+				.uid("Deserialization");
+
+		DataStream<Measurement> sortedStream = sourceStream
+				.keyBy(Measurement::getSensorId)
+				.process(new SortMeasurementFunction())
+				.name("Sorting")
+				.uid("Sorting");
+
+		KeyedStream<Measurement, Integer> keyedSortedStream =
+				DataStreamUtils.reinterpretAsKeyedStream(
+						sortedStream,
+						Measurement::getSensorId);
+
+		DataStream<WindowedMeasurements> aggregatedPerLocation = keyedSortedStream
+				.window(SlidingEventTimeWindows.of(Time.of(1, TimeUnit.MINUTES), Time.of(1, TimeUnit.SECONDS)))
+				.aggregate(new MeasurementWindowAggregatingFunction(),
+						new MeasurementWindowProcessFunction())
+				.name("WindowedAggregationPerLocation")
+				.uid("WindowedAggregationPerLocation");
+
+		if (isLocal(parameters)) {
+			aggregatedPerLocation.print()
+					.name("NormalOutput")
+					.uid("NormalOutput")
+					.disableChaining();
+		} else {
+			aggregatedPerLocation.addSink(new DiscardingSink<>())
+					.name("NormalOutput")
+					.uid("NormalOutput")
+					.disableChaining();
+		}
+
+		env.execute(CheckpointingJobSolution3.class.getSimpleName());
+	}
+
+	public static class SortMeasurementFunction
+			extends KeyedProcessFunction<Integer, Measurement, Measurement> {
+
+		private MapState<Long, List<Measurement>> mapState;
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
+			super.open(parameters);
+
+			MapStateDescriptor<Long, List<Measurement>> desc =
+					new MapStateDescriptor<>(
+							"events",
+							Types.LONG,
+							Types.LIST(Types.POJO(Measurement.class))
+					);
+			mapState = getRuntimeContext().getMapState(desc);
+		}
+
+		@Override
+		public void processElement(Measurement value, Context ctx, Collector<Measurement> out) throws Exception {
+			TimerService timerService = ctx.timerService();
+			Long currentTimestamp = ctx.timestamp();
+
+			if (currentTimestamp > timerService.currentWatermark()) {
+				List<Measurement> measurementList = mapState.get(currentTimestamp);
+				if (measurementList == null) {
+					measurementList = new ArrayList<>();
+				}
+				measurementList.add(value);
+				mapState.put(currentTimestamp, measurementList);
+				timerService.registerEventTimeTimer(currentTimestamp);
+			}
+		}
+
+		@Override
+		public void onTimer(long timestamp, OnTimerContext ctx,
+							Collector<Measurement> out) throws Exception {
+			List<Measurement> measurementList = mapState.get(timestamp);
+			for (Measurement measurement : measurementList) {
+				out.collect(measurement);
+			}
+			mapState.remove(timestamp);
+		}
+	}
+
+	public static class MeasurementWindowAggregatingFunction implements
```

Comment:
Please add a javadoc that explains at least the assumptions that you can make here, i.e. that events are sorted (on a per-key basis).

Reply:
done.